juju/model.py

```diff
@@ -134,22 +103,13 @@ def __len__(self):
     def __iter__(self):
         return iter(self._data)
 
-    def __getitem__(self, key):
-        return self._data[key]
-
-
-class RelationList(LazyList):
-    def __init__(self, relation_name, local_unit, backend, cache):
-        self.relation_name = relation_name
-        self._local_unit = local_unit
-        self._backend = backend
-        self._cache = cache
-
-    def _load(self):
-        relations = []
-        for relation_id in self._backend.relation_ids(self.relation_name):
-            relations.append(Relation(self.relation_name, relation_id, self._local_unit, self._backend, self._cache))
-        return relations
+    def __getitem__(self, relation_name):
+        relation_list = self._data[relation_name]
+        if relation_list is None:
+            relation_list = self._data[relation_name] = []
+            for relation_id in self._backend.relation_ids(relation_name):
+                relation_list.append(Relation(relation_name, relation_id, self._local_unit, self._backend, self._cache))
+        return list(relation_list)  # Return a copy so our copy cannot be modified.
```

Comment:
I think we shouldn't do that yet. If we teach people that we always returns copies, they'll learn that they can poke at the model data without consequences, which means we'll really need to be careful about what data we hand out, and if we look around the current model code, there already other cases where we don't return copies, not to mention all the model instances themselves are not copies. Also, this means the cost of a read becomes a write, which tends not to be a great tradeoff. So at least right now it seems better to try the opposite: let's teach that fiddling with the model data is not okay. If we see people fiddling with our data in the wild, we may actually go overboard and try harder to make all of it immutable.